### PR TITLE
core: make clist singly-linked

### PR DESCRIPTION
**Pull request by kaspar030**
clist was doubly linked in order to enable removal of threads from a run queue in constant time.

Today I realized that only the currently running thread can remove itself from a running queue (by blocking for a msg, mutex, sleeping, ...), and a singly linked but circular list offers O(1) removal of the head element.
So I changed clist accordingly.

This carves off 4 byte of tcb and increases context switch ~~time~~ performance by ~4%.

~~Waiting for #4557.~~

### CODE REVIEW

**Comment by OlegHahm** [comment]
Nice!

**Comment by haukepetersen** [comment]
I like!

File: tests/unittests/tests-core/tests-core-clist.c [hunk 1]
```diff
@@ -16,7 +16,8 @@
 
 #define TEST_CLIST_LEN    (8)
 
-static clist_node_t tests_clist_buf[TEST_CLIST_LEN];
+static list_node_t tests_clist_buf[TEST_CLIST_LEN];
```
Comment (authmillenon) [inline tests/unittests/tests-core/tests-core-clist.c]: Here and below: the datatype is called `list` now so the variable, function and file names should reflect that.

Comment (kaspar030) [inline tests/unittests/tests-core/tests-core-clist.c]: Even if clist_*() works with list_node_t objects?

Comment (authmillenon) [inline tests/unittests/tests-core/tests-core-clist.c]: But doesn't `clist` stand for circular list? (or is it that still)?

Comment (kaspar030) [inline tests/unittests/tests-core/tests-core-clist.c]: Yes. I shared the list node definition between the two header files. Now I copied the definition, do you prefer it that way?

**Comment by kaspar030** [comment]
- rebased
- made clist.h independent from list.h

File: tests/unittests/tests-core/tests-core-clist.c (hunk 1 above)
Comment (authmillenon) [inline tests/unittests/tests-core/tests-core-clist.c]: Why copy when you can 

```c
typedef list_t clist_t;
```

Comment (kaspar030) [inline tests/unittests/tests-core/tests-core-clist.c]: typedef'ed it is now.

**Comment by authmillenon** [comment]
Why does this depend on #4557? This data structure should be independent of that change, right?

**Comment by kaspar030** [comment]
> Why does this depend on #4557? This data structure should be independent of that change, right?

#4557 introduces link.h. With copying the struct, this PR was independent. With the typedef, it's not.

**Comment by authmillenon** [comment]
Why can't you take the list.h introduction over here and make #4557 dependent on this PR. That seems to be the more logical dependency graph to me.

**Comment by kaspar030** [comment]
The perfect graph would be a "list.h" introduction PR, and then #4557 and this. Moving the list.h introduction here just shuffles the imperfect graph solution here.

I'd prefer merging #4557, then merging this.

**Comment by OlegHahm** [comment]
Dependency not yet merged. Postponed.

**Comment by kaspar030** [comment]
Sad.

**Comment by OlegHahm** [comment]
Agreed, but currently I don't see realistic chances for #4557 to get merged until tomorrow. Do you disagree?

**Comment by kaspar030** [comment]
Can't review it myself...

**Comment by OlegHahm** [comment]
You can kick your reviewers. ;)

File: core/include/clist.h [hunk 2]
```diff
@@ -13,6 +14,14 @@
  * @file
  * @brief       Circular linked list
  *
+ * This file contains a circularly linked list implementation.
+ *
+ * clist_insert(), clist_remove_head() and clist_advance() take constant time.
+ *
+ * Each list is represented as a "clist_node_t". It's only member, the "next"
+ * pointer, points to the last entry in the list, whose "next" pointer points to
+ * the first entry.
```
Comment (haukepetersen) [inline core/include/clist.h]: Something reads weird here, how about 
```
... . It's only member, the "next" pointer, points to the next entry in the list. The last element's "next" pointer points to the first entry.
```
￼

Comment (haukepetersen) [inline core/include/clist.h]: missread the sentence, never mind...

**Comment by authmillenon** [comment]
Needs rebase.

**Comment by kaspar030** [comment]
- rebased, squashed

**Comment by haukepetersen** [comment]
Code looks good, works as expected (on 'native' and some ARM boards), code size is smaller than before -> ACK!

**Comment by cgundogan** [comment]
code looks sensible and changes make sense. ACK

**Pull request merged**